webhook/webhook.go

```diff
@@ -0,0 +1,90 @@
+package webhook
+
+import (
+  "crypto/hmac"
+  "crypto/sha256"
+  "encoding/hex"
+  "time"
+  "net/http"
+  "fmt"
+  "errors"
+  "encoding/json"
+  "bytes"
+  "encoding/gob"
+
+  log "github.com/gophish/gophish/logger"
+)
+
+const (
+  DefaultTimeoutSeconds = 10
+  MinHttpStatusErrorCode = 400
+)
+
+//TODO rename to Sender because "http" contains Transport too
+type Transport struct {
+  Client *http.Client
+}
+
+//TODO
+
+
+func (whTr *Transport) Send(url string, secret string, data interface{}) error {
+  if whTr.Client == nil {
+    errMsg := "Client must be initialized"
+    log.Error(errMsg)
+    panic(errMsg)
+  }
+  jsonData, err := json.Marshal(data)
+  if err != nil {
+    log.Error(err)
+    return err
+  }
+  req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
+  ts := int32(time.Now().Unix())
+  signat, err := sign(secret, data, ts)
+  req.Header.Set("X-Gophish-Signature", signat)
+  req.Header.Set("Content-Type", "application/json")
+  resp, err := whTr.Client.Do(req)
+  if err != nil {
+    log.Error(err)
+    return err
+  }
+  defer resp.Body.Close()
+
+
+  //TODO
+  if resp.StatusCode >= MinHttpStatusErrorCode {
+    errMsg := fmt.Sprintf("http status of response: %s", resp.Status)
+    log.Error(errMsg)
+    return errors.New(errMsg)
+  }
+  return nil
+}
+
+
+//TODO
+func sign(secret string, data interface{}, ts int32) (string, error) {
```

Comment:
Okay, I've removed a timestamp.

We send body in JSON to a remote server. That is, `json.Marshal(data/payload)` will have to be called anyway, either in `SendWebhook` or anywhere in other methods which call `SendWebhook`. How do you say we can avoid a redundant conversion to JSON?

Reply:
I think we should just call it in `Send`, and then pass that data (right now this is `jsonData`) to the `sign()` function as a `[]byte`. This way, we only marshal once for the webhook, and it makes the `sign` function a bit more focused, since it's really there to sign bytes.